lldb/source/Plugins/TypeSystem/Swift/SwiftDemangle.h

```diff
@@ -70,7 +70,82 @@ static swift::Demangle::NodePointer
 GetDemangledType(swift::Demangle::Demangler &dem, llvm::StringRef name) {
   return GetType(dem.demangleSymbol(name));
 }
+
+/// Shared ownership wrapper for swift::Demangle::NodePointer
+///
+/// Can be valid or invalid. A valid instance stores a non-null NodePointer and
+/// a shared pointer to swift::Demangle::Demangler that owns this NodePointer.
+/// An invalid instance has NodePointer equal to nullptr.
+class SharedDemangledNode {
```

Comment:
I wonder what the tradeoffs are between having a `std::shared_ptr<swift::Demangle::Demangler>` inside the class, instead of having users store a `std::shared_ptr<SharedDemangledNode>`.

Reply:
We need to be able to reference the same Demangler from two instances of SharedDemangledNode. For example, GetType returns a new instance whose NodePointer is pointing to a subtree of the current node, and its memory is managed by the current Demangler.